src/com/amazon/ion/impl/IonReaderContinuableApplicationBinary.java

```diff
@@ -0,0 +1,689 @@
+import static com.amazon.ion.SystemSymbols.MAX_ID_SID;
+import static com.amazon.ion.SystemSymbols.NAME_SID;
+import static com.amazon.ion.SystemSymbols.SYMBOLS_SID;
+import static com.amazon.ion.SystemSymbols.VERSION_SID;
+
+/**
+ * An IonCursor capable of application-level parsing of binary Ion streams.
+ */
+class IonReaderContinuableApplicationBinary extends IonReaderContinuableCoreBinary implements IonReaderContinuableApplication {
+
+    // The UTF-8 encoded bytes representing the text `$ion_symbol_table`.
+    private static final byte[] ION_SYMBOL_TABLE_UTF8;
+
+    static {
+        ION_SYMBOL_TABLE_UTF8 = SystemSymbols.ION_SYMBOL_TABLE.getBytes(StandardCharsets.UTF_8);
+    }
+
+    // An IonCatalog containing zero shared symbol tables.
+    private static final IonCatalog EMPTY_CATALOG = new SimpleCatalog();
+
+    // Initial capacity of the ArrayList used to hold the text in the current symbol table.
+    private static final int SYMBOLS_LIST_INITIAL_CAPACITY = 128;
+
+    // The imports for Ion 1.0 data with no shared user imports.
+    private static final LocalSymbolTableImports ION_1_0_IMPORTS
+        = new LocalSymbolTableImports(SharedSymbolTable.getSystemSymbolTable(1));
+
+    // The text representations of the symbol table that is currently in scope, indexed by symbol ID. If the element at
+    // a particular index is null, that symbol has unknown text.
+    private String[] symbols;
+
+    // The max ID of the local symbol table.
+    private int localSymbolMaxId = -1;
+
+    // The catalog used by the reader to resolve shared symbol table imports.
+    private final IonCatalog catalog;
+
+    // Uses the underlying raw reader to read the symbol tables from the stream.
+    private final SymbolTableReader symbolTableReader;
+
+    // The shared symbol tables imported by the local symbol table that is currently in scope.
+    private LocalSymbolTableImports imports = ION_1_0_IMPORTS;
+
+    // The first lowest local symbol ID in the symbol table.
+    private int firstLocalSymbolId = imports.getMaxId() + 1;
+
+    // ------
+
+    /**
+     * Constructs a new reader from the given byte array.
+     * @param builder the builder containing the configuration for the new reader.
+     * @param bytes the byte array containing the bytes to read.
+     * @param offset the offset into the byte array at which the first byte of Ion data begins.
+     * @param length the number of bytes to be read from the byte array.
+     */
+    IonReaderContinuableApplicationBinary(IonReaderBuilder builder, byte[] bytes, int offset, int length) {
+        super(builder.getBufferConfiguration(), bytes, offset, length);
+        this.catalog = builder.getCatalog() == null ? EMPTY_CATALOG : builder.getCatalog();
+        symbols = new String[SYMBOLS_LIST_INITIAL_CAPACITY];
+        symbolTableReader = new SymbolTableReader();
+        resetImports();
+        registerIvmNotificationConsumer((x, y) -> {
```

Comment:
Curious why this is a callback for this specialization instead of an override.

Reply:
I think you're asking why we inject the logic using `registerIvmNotificationConsumer` rather than just having a `handleIvm` method that is optionally overridden by subclasses.

The answer is that we envision use cases where any of these IonCursor implementations can be used as delegates from wrappers that do not override them directly. The first example is `IonReaderNonContinuableSystem` (see #514), which delegates to an injected IonReaderContinuableCore provided via its constructor. The IonReaderNonContinuableSystem has to inject its custom IVM handling logic even though its IonReaderContinuableCore (which may be backed by either binary or text) has already been instantiated.